src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                for (var i = 0; i < invocationArguments.ToArray().Length; i++)
+                {
+                    var semanticFacts = document.GetRequiredLanguageService<ISemanticFactsService>();
+                    var argumentParameter = semanticFacts.FindParameterForArgument(invocationSemanticModel, invocationArguments.ToArray()[i], cancellationToken);
+                    if (argumentParameter.Equals(mappedParameter, SymbolEqualityComparer.Default))
+                    {
+                        var updatedInvocationArgument = updatedInvocationArguments.ToArray()[i];
+                        var argumentExpression = syntaxFacts.GetExpressionOfArgument(updatedInvocationArgument);
+                        var parenthesizedArgumentExpression = generator.AddParentheses(argumentExpression, includeElasticTrivia: false);
+                        expression = expression.ReplaceNode(oldNode, parenthesizedArgumentExpression);
+                        parameterMapped = true;
+                        break;
+                    }
+                }
+
+                // This is special cased for optional parameters: if the invocation does not have an argument 
+                // corresponding to the optional parameter, then it generates an expression from the optional parameter.
+                if (mappedParameter.HasExplicitDefaultValue && !parameterMapped)
+                {
+                    parameterIsNamed = true;
+                    var generatedExpression = GenerateExpressionFromOptionalParameter(mappedParameter);
+                    var parenthesizedGeneratedExpression = generator.AddParentheses(generatedExpression, includeElasticTrivia: false);
+                    expression = expression.ReplaceNode(oldNode, parenthesizedGeneratedExpression);
+                }
+            }
+
+            return (expression, parameterIsNamed);
+        }
+
+        /// <summary>
+        /// This method goes through all the invocation sites and adds a new argument with the expression to be added.
+        /// It also introduces a parameter at the original method site.
+        /// 
+        /// Example:
+        /// public void M(int x, int y)
+        /// {
+        ///     int f = x * y; // highlight this expression
+        /// }
+        /// 
+        /// public void M1()
+        /// {
+        ///     M(5, 6);
+        /// }
+        /// ---------------------------------------------------->
+        /// 
+        /// public void M(int x, int y, int f)
+        /// {
+        /// }
+        /// 
+        /// public void M1()
+        /// {
+        ///     M(5, 6, 5 * 6);
+        /// }
+        /// </summary>
+        private async Task<Solution> ModifyDocumentInvocationsAndIntroduceParameterAsync(Compilation compilation,
+            Document originalDocument, Document document, Solution modifiedSolution, IMethodSymbol methodSymbol,
+            SyntaxNode containingMethod, TExpressionSyntax expressionCopy, TExpressionSyntax expression,
```

Comment:
why do you need an expressionCopy.  tihs still remains unexplained adquately.

Reply:
don't anymore